### DISCUSSION TITLE
Automatic retry with HTTP 4xx errors

### DISCUSSION THREAD

**Discussion by ttsirkia** [trpc/trpc#2263] [General]
With the default React Query settings, it will try failed responses three times before setting the query to error state. This also happens with HTTP responses 4xx which are not likely to be temporarily issues which a 500 could be.

As React Query is not bound to HTTP, it is mentioned that it cannot make any difference between HTTP errors. Should tRPC do and change the default retry logic to retry only, for example, with 5xx responses?

**Comment by ttsirkia** [comment]
The main reason is that, for example, in case with Forbidden response, it takes multiple seconds to show the error as it increases the delay between the retry requests.

**Comment by iway1** [comment]
It’s pretty easy to do this yourself in the query client. react query provides a mechanism for easily adjusting the retry time, and writing custom logic for deciding whether a retry should occur. 

It’s really valuable that trpc just uses react query’s default behavior, it makes it seamless to move between vanilla react query and trpc

**Reply by ttsirkia** [reply]
Yes, I know that it is rather easy. I was just thinking could trpc provide here some additional value as it is purely http-based library whereas React Query is not. It would make it easier for developers if the library provided better defaults to avoid eight second waits after ACCESS_FORBIDDEN errors, for example.

**Reply by KATT** [reply]
@ttsirkia please make a PR to update [this code](https://github.com/trpc/trpc/blob/0f9406826c97c159e67a9c6eacbeaf4150e6ec56/examples/next-prisma-starter/src/pages/_app.tsx#L72-L75) to something more useful if you want 🙏

**Reply by ttsirkia** [reply]
I’ll take a look at this! Maybe an option which would change the default logic would be the best DX. Easy to enable but doesn’t change the default behavior if you don’t want to.

**Reply by jlalmes** [reply]
Related discussion: https://github.com/trpc/trpc/issues/1714
